server/plan_replayer_test.go

```diff
@@ -242,3 +251,199 @@ func forEachFileInZipBytes(t *testing.T, b []byte, fn func(file *zip.File)) {
 		fn(f)
 	}
 }
+
+func fetchZipFromPlanReplayerAPI(t *testing.T, client *testServerClient, filename string) *zip.Reader {
+	resp0, err := client.fetchStatus(filepath.Join("/plan_replayer/dump/", filename))
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, resp0.Body.Close())
+	}()
+	body, err := io.ReadAll(resp0.Body)
+	require.NoError(t, err)
+	b := bytes.NewReader(body)
+	z, err := zip.NewReader(b, int64(len(body)))
+	require.NoError(t, err)
+	return z
+}
+
+func getInfoFromPlanReplayerZip(
+	t *testing.T,
+	z *zip.Reader,
+) (
+	jsonTbls []*handle.JSONTable,
+	metas []map[string]string,
+	errMsgs []string,
+) {
+	for _, zipFile := range z.File {
+		if strings.HasPrefix(zipFile.Name, "stats/") {
+			jsonTbl := &handle.JSONTable{}
+			r, err := zipFile.Open()
+			require.NoError(t, err)
+			//nolint: all_revive
+			defer func() {
+				require.NoError(t, r.Close())
+			}()
+			buf := new(bytes.Buffer)
+			_, err = buf.ReadFrom(r)
+			require.NoError(t, err)
+			err = json.Unmarshal(buf.Bytes(), jsonTbl)
+			require.NoError(t, err)
+
+			jsonTbls = append(jsonTbls, jsonTbl)
+		} else if zipFile.Name == "sql_meta.toml" {
+			meta := make(map[string]string)
+			r, err := zipFile.Open()
+			require.NoError(t, err)
+			//nolint: all_revive
+			defer func() {
+				require.NoError(t, r.Close())
+			}()
+			_, err = toml.NewDecoder(r).Decode(&meta)
+			require.NoError(t, err)
+
+			metas = append(metas, meta)
+		} else if zipFile.Name == "errors.txt" {
+			r, err := zipFile.Open()
+			require.NoError(t, err)
+			//nolint: all_revive
+			defer func() {
+				require.NoError(t, r.Close())
+			}()
+			content, err := io.ReadAll(r)
+			require.NoError(t, err)
+			errMsgs = strings.Split(string(content), "\n")
+		}
+	}
+	return
+}
+
+func TestDumpPlanReplayerAPIWithHistoryStats(t *testing.T) {
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/domain/sendHistoricalStats", "return(true)"))
+	defer func() {
+		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/domain/sendHistoricalStats"))
+	}()
```

Comment:
I recommend using the context to inject the failure instead of the `fail-point`.
Here is an example https://github.com/pingcap/tidb/blob/master/planner/core/plan_cache.go#L48.
Using the context seems more go-style and easier to use (don't need to execute 'enable failpoint' to modify all source code...)

Reply:
The logic related to the failpoint is irrelevant to the content of this PR. Just that logic is needed to make the test here run as expected.
So I tend not to modify that in this PR.